subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/WatchingVirtualFileSystem.java

```diff
@@ -163,11 +174,15 @@ public void updateMustWatchDirectories(Collection<File> mustWatchDirectories) {
     @Override
     public void beforeBuildFinished(boolean watchingEnabled) {
         if (watchingEnabled) {
-            handleWatcherRegistryEvents("for current build");
-            buildRunning = false;
-            producedByCurrentBuild.set(DefaultFileHierarchySet.of());
-            printStatistics("retains", "till next build");
-            updateWatchRegistry(watchRegistry -> {}, this::invalidateAll);
+            getRoot().update(currentRoot -> {
+                buildRunning = false;
+                producedByCurrentBuild.set(DefaultFileHierarchySet.of());
+                SnapshotHierarchy newRoot = watchRegistry == null
+                    ? currentRoot.empty()
+                    : handleWatcherRegistryEvents(currentRoot, watchRegistry, "for current build");
```

Comment:
It might make sense to extract this to a method.

Reply:
I moved the null check logic to `handleWatcherRegistryEvents`.